New feature: Go to point

alvarocasadoc opened the pull request:
Hi everyone! 
I am opening a PR that implements a "go to point" function in the editor as mentioned in issue #596.

### Changes

1. I have added an extra "zoom" action (visible from the toolbar and the view menu):

<img width="411" alt="image" src="https://user-images.githubusercontent.com/7248474/215321315-357346e6-1bf0-48aa-a172-88f70b165c28.png">

<img width="275" alt="image" src="https://user-images.githubusercontent.com/7248474/215321347-98d7a483-cf7e-46f0-a605-40c5cbcf79b6.png">

2. You can find the new icons inside `src/libs/vmisc/share/resources/icons/win.icon.theme`:
    - 16x16: ![zoom_to_point](https://user-images.githubusercontent.com/7248474/215321532-dfe9707f-4acc-4024-805f-8e0b9cf87db4.png)

    - 24x24: ![zoom_to_point](https://user-images.githubusercontent.com/7248474/215321558-cf5a1dc9-9459-4ca6-ad44-f1256a4106ca.png)

    - 32x32: ![zoom_to_point](https://user-images.githubusercontent.com/7248474/215321566-86d39538-3a7c-4e98-871c-b55d7856b51a.png)

3. There is a new `MainWindow::zoomToPoint()` function that implements the following workflow:
    - Show a dialog with a dropdown with all the points in alphabetical order.
    - Once selected, the graphics view focuses on the point


I have attached a video where you can see how it works. Please, let me know if you have any feedback in case you would like to change the UI/UX workflow, the icons, or anything else :)

https://user-images.githubusercontent.com/7248474/215321714-dc68f2da-f94d-4bca-83c1-8a0de1d707e9.mov

alvarocasadoc committed:
Added zoom to point action

alvarocasadoc committed:
Added icons for zoom to point action

alvarocasadoc committed:
Added zoom to point action to View menu

alvarocasadoc committed:
Removed debug message in MainWindow::zoomToPoint()

DSCaskey commented:
Hmmm... some issues

1) The 16x16 and 32x32 icons don't belong. - not the same as the rest of the icons.

2) The use of the shortcut Ctrl + P should not be used as it's supposed to be used for the File->Print menu item. 

3) The shortcut (whatever it is) needs to be added to the Help->Shortcuts dialog. 

4) The tooltip for the toolbar item lacks the shortcut.

5) The toolbar zoomToPoint action should be disabled (greyed out) when in Piece & Layout mode as it's not an option in those modes.

6) ***Big issue here... the point names in the combobox are showing up multiple times... ?

![topoint2](https://user-images.githubusercontent.com/31944718/215397597-2190e008-e8bd-4e03-87de-712882d2056d.png)

Hint: You've made the false assumption that pattern->DataGObjects() contains only 1 copy of each point. 
Hint 2:  setDuplicatesEnabled(bool enable)

 And some requests.
 
 7) I'd like to see the ? in the Window title bar gone. It's not used and is just plain annoying. 
  
![topoint](https://user-images.githubusercontent.com/31944718/215398125-7ebb8b85-930c-455e-803b-e839f8f792e0.png)

It can be removed... search through the code, and see how I did this...

![topoint3](https://user-images.githubusercontent.com/31944718/215398253-bb44ed74-78f9-4f1b-9fcf-1fb48afd3305.png)

8) Not a big deal, but I'd like to see any new methods @briefed (in the cpp) so they will show up in documentation - such as running Doxygen. 

9) Again not a big deal, but here QString item ... QString name would be clearer. Don't want to have to think what "item" is... "name" tells me that by simply reading. That also would allow the use of "item" as the iterator name...which is what you are iterating - combobox items.  "it" is not a good name for the iterator... either use the common i, j, k or something descriptive.  

This is easier to read:

if (item.value()->name() == name)

10) Try to keep the line length of code <120

DSCaskey commented:
if (qApp->getCurrentScene() != draftScene) return;

Is not needed if the action is properly disabled for Piece & Layout mode as you wouldn't even get here.

DSCaskey commented:
This would also require an Lupdate as the UI has changed.

DSCaskey commented:
After playing a bit more... Several other observations. 

IMO Selecting the tool button, select combobox, select a point from a combobox and then having to select OK (4 clicks) AND given the fact the combobox in the dialog is set to not accept typed input to (search) select a point - it's a bit tedious.  I would have just made the toolbar tool button action a drop down. - 1 click, select and let go.  Like in the Toolbox toolbar. 

If you could type a point name in the combobox in the dialog I'd keep that, tied to a shortcut and the View menu. 

Then you'd have the best of both worlds. A quick way to find a point,  with a way to use a shortcut and /or search rather than selecting a point name from a list.

DSCaskey commented:
Here's the blank zoom svg...

![blank_zoom](https://user-images.githubusercontent.com/31944718/215643520-f437a753-45f1-4cd3-b651-c8d7f078f17d.svg)

slspencer commented:
Hi @alvarocasadoc 

Thanks for the great work. 
The Zoom to Point tool is needed, and you've done a great job with it.

There were a few comments, I had to review it a few times to understand what was being requested, so here's a summary. If you need help with any of these let us know.

Required:
- [ ] Update the 16x16 and 32x32 icons to match the 24x24 icon.
- [ ] Update the shortcut to Ctrl + Alt + P or Ctrl + Shift + P
- [ ] Add this new shortcut to Help -> Shortcuts dialog
- [ ] Add a tooltip to the zoomToPoint icon in the toolbar
- [ ] Prevent duplicate point names in the combobox  
- [ ] Disable the tooltip for Piece mode and Layout mode.   
   ```if (qApp->getCurrentScene() != draftScene) return;``` can be removed if zoomToPoint when disabled in Piece & Layout modes


Optional:
- [ ] Suppress the ? in the Window title bar
- [ ] Add a @brief in the cpp to enable Doxygen documentation
- [ ] Rename QString item to QString name &  use "item" as the iterator name or the common i, j, k 
- [ ] Try to keep the line length of code <120
- [ ] Enhance the toolbar icon with a dropdown box for quick use

DSCaskey commented:
>Update the 16x16 and 32x32 icons to match the 24x24 icon.

I provided the base icon svg in the PR comments.

>Disable the tooltip for Piece mode and Layout mode.
if (qApp->getCurrentScene() != draftScene) return; can be removed if
zoomToPoint when disabled in Piece & Layout modes

The whole toolbutton needs to be disabled in Piece and Layout Mode not just
the tooltip. If the button is disabled there is no need to check if it's in
draft mode as it's redundant.


On Mon, Jan 30, 2023, 9:24 PM slspencer ***@***.***> wrote:

> Hi @alvarocasadoc <https://github.com/alvarocasadoc>
>
> Thanks for the great work.
> The Zoom to Point tool is needed, and you've done a great job with it.
>
> There were a few comments, I had to review it a few times to understand
> what was being requested, so here's a summary. If you need help with any of
> these let us know.
>
> Required:
>
>    - Update the 16x16 and 32x32 icons to match the 24x24 icon.
>    - Update the shortcut to Ctrl + Alt + P or Ctrl + Shift + P
>    - Add this new shortcut to Help -> Shortcuts dialog
>    - Add a tooltip to the zoomToPoint icon in the toolbar
>    - Prevent duplicate point names in the combobox
>    - Disable the tooltip for Piece mode and Layout mode.
>    if (qApp->getCurrentScene() != draftScene) return; can be removed if
>    zoomToPoint when disabled in Piece & Layout modes
>
> Optional:
>
>    - Suppress the ? in the Window title bar
>    - Add a @brief <https://github.com/brief> in the cpp to enable Doxygen
>    documentation
>    - Rename QString item to QString name & use "item" as the iterator
>    name or the common i, j, k
>    - Try to keep the line length of code <120
>    - Enhance the tooltip icon with a dropdown box for quick use
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/844#issuecomment-1409649324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHTXADUAW2WG2PTQZL5LPMDWVBZVHANCNFSM6AAAAAAUKFYRC4>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

alvarocasadoc committed:
Updated zoom to point icons

alvarocasadoc committed:
Added zoom to point shortcut to Help>Shortcuts dialog

alvarocasadoc committed:
Improved zoom to point feature

- Changed shortcut to Ctrl + Alt + P
- Added the shortcut to the tooltip of the zoom to point action in the toolbar
- Added a combobox for the zoom to point feature
- Supressed the help hint button ("?") in the zoom to point dialog
- Disabled the zoom to point button and combo box in Piece and Layout modes
- Improved naming of variables

alvarocasadoc commented:
Hello everyone!
Thank you so much for all your feedback, and thank you @slspencer for the summary.

I have just pushed my latest changes. Here is the current status:

Required:

- [x] Updated the 16x16 and 32x32 icons to match the 24x24 icon. I have replaced the 3 files using the svg file provided by @DSCaskey. Please let me know if I should make any other modification.
- [x] Updated the shortcut to Ctrl + Alt + P.
- [x] Added the new shortcut to Help -> Shortcuts dialog.
- [x] Added the shortcut to the tooltip of the zoomToPoint icon in the toolbar.
- [ ] Prevent duplicate point names in the combobox. I could not verify this on my machine. I will try to test it on Windows as soon as I can. 
- [x] Disabled the zoom to point button for Piece mode and Layout mode. Removed `if (qApp->getCurrentScene() != draftScene) return;`

Optional:

- [ ] Suppressed the ? in the Window title bar. I have added new window flags to the dialog, but I cannot test it since I don't have access to a Windows machine right now. 
- [x] Added @brief in the cpp to enable Doxygen documentation
- [x] Renamed QString item to QString pointName & use "item" as the iterator name. 
- [x] Tried to keep the line length of code <120
- [x] Added a dropdown box to the toolbar for quick use. So far, I have kept both interaction flows (button -> dialog and combobox).

I will update you when I manage to fix the remaining issues.

I also attached a short video of the new combobox vs the button + dialog interaction. 
Let me know your thoughts! 😄 


https://user-images.githubusercontent.com/7248474/215898123-3f2a9ba6-caa5-4280-99b8-c29dfae1f3cb.mov

DSCaskey commented:
I'll check it out later when I get home from work.

Icons look good. :)

Just another thought I had... while you can disable duplicates in a
combobox,  you can also use removeDuplicates() for the QStringList after
populating it before passing it to a combobox.

On Tue, Jan 31, 2023, 5:30 PM Álvaro ***@***.***> wrote:

> Hello everyone!
> Thank you so much for all your feedback, and thank you @slspencer
> <https://github.com/slspencer> for the summary.
>
> I have just pushed my latest changes. Here is the current status:
>
> Required:
>
>    - Updated the 16x16 and 32x32 icons to match the 24x24 icon. I have
>    replaced the 3 files using the svg file provided by @DSCaskey
>    <https://github.com/DSCaskey>. Please let me know if I should make any
>    other modification.
>    - Updated the shortcut to Ctrl + Alt + P.
>    - Added the new shortcut to Help -> Shortcuts dialog.
>    - Added the shortcut to the tooltip of the zoomToPoint icon in the
>    toolbar.
>    - Prevent duplicate point names in the combobox. I could not verify
>    this on my machine. I will try to test it on Windows as soon as I can.
>    - Disabled the zoom to point button for Piece mode and Layout mode.
>    Removed if (qApp->getCurrentScene() != draftScene) return;
>
> Optional:
>
>    - Suppressed the ? in the Window title bar. I have added new window
>    flags to the dialog, but I cannot test it since I don't have access to a
>    Windows machine right now.
>    - Added @brief <https://github.com/brief> in the cpp to enable Doxygen
>    documentation
>    - Renamed QString item to QString pointName & use "item" as the
>    iterator name.
>    - Tried to keep the line length of code <120
>    - Added a dropdown box to the toolbar for quick use. So far, I have
>    kept both interaction flows (button -> dialog and combobox).
>
> I will update you when I manage to fix the remaining issues.
>
> I also attached a short video of the new combobox vs the button + dialog
> interaction.
> Let me know your thoughts! 😄
>
>
> https://user-images.githubusercontent.com/7248474/215898123-3f2a9ba6-caa5-4280-99b8-c29dfae1f3cb.mov
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/844#issuecomment-1411156365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHTXADVQMVHJBSTTGDQDAI3WVGG7FANCNFSM6AAAAAAUKFYRC4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DSCaskey reviewed (changes requested):
If the Input dialog combobox is set to enabled=true a user can type and find a point name. The box will move the index to match what is typed.  Since it's a throw away list we don't care that any name not in the list is added... just won't zoom to anything.

![dialog_combo](https://user-images.githubusercontent.com/31944718/215955152-5628c483-a4ac-4eb8-ab0a-e1f572e9fec5.png)


The routine updateZoomToPointComboBox() does not belong in the updateViewToolbar(). The combobox is located in the ZoomToolbar. Makes no sense to have a change in the View preference settings updating the comboBox. 

May be a moot point anyway. There's an issue just populating the zoomPointComboBox, and not accounting for the fact the point names can change. Renamed P3 to Point3, but the combobox is not updated. 

![nochange](https://user-images.githubusercontent.com/31944718/215955722-d5e04ae6-29e1-4c8e-9965-d8f13983e742.png)

DSCaskey reviewed (changes requested):
Slight glitch in the alignment. 

![shortcuts](https://user-images.githubusercontent.com/31944718/215956247-0b8d2479-6721-4609-8912-3dc7059eb34c.png)

DSCaskey reviewed (changes requested):
If the Input dialog combobox is set to enabled=true a user can type and find a point name. The box will move the index to match what is typed.  Since it's a throw away list we don't care that any name not in the list is added... just won't zoom to anything.

![dialog_combo](https://user-images.githubusercontent.com/31944718/215955152-5628c483-a4ac-4eb8-ab0a-e1f572e9fec5.png)


The routine updateZoomToPointComboBox() does not belong in the updateViewToolbar(). The combobox is located in the ZoomToolbar. Makes no sense to have a change in the View preference settings updating the comboBox. Not sure if the same call belongs in the ClosedDialogWithApply() routine either. 

Also the zoomToPointComboBox->blockSignals() can go inside the update routine so you don't have to implicitly call those every time before and after the update routine. 

May be a moot point anyway. There's an issue just populating the zoomPointComboBox, and not accounting for the fact the point names can change. 

![nochange](https://user-images.githubusercontent.com/31944718/215955722-d5e04ae6-29e1-4c8e-9965-d8f13983e742.png)

Probably makes more sense to connect any change in the pattern data to update the point combobox, and those 2 calls would not be neccessary.

alvarocasadoc commented:
- I like your idea of making the combobox editable. I will upload the changes!
- I am calling `updateZoomToPointComboBox()` inside `updateViewToolbar()` and `ClosedDialogWithApply()` because its the way I found for updating the list of points when the user changes mode to/from Piece/Layout and when a new point or line is added. **I think I can move the first from `updateViewToolbar()` to `showDraftMode()` since it makes more sense.**
- I agree with including `zoomToPointComboBox->blockSignals()` inside the update routine.
- Thanks for spotting the issue with the list not being updated when a point name is changed. I will take a look at it!
- I will take care of the alignment problems in the Help>Shortcuts dialog.

alvarocasadoc committed:
Moved blockSignals inside MainWindow::updateZoomToPointComboBox

alvarocasadoc committed:
Moved updateZoomToPointComboBox call to showDraftMode

DSCaskey commented:
>I am calling updateZoomToPointComboBox() inside updateViewToolbar()

There is no need to... updateViewToolbar() (not to be confused with the Zoom Toolbar where you put the combobox) is there to update the toolbuttons such as show seam allowance if the Preferences were changed. Has nothing to do with the points, draftblock, or otherwise. Trace back the routine to the connect statement and you will find it's called when the settings data has changed. Besides, you're only updating the combobox,  not the toolbar. Again... If it were me I would connect a change in the doc (pattern) to update the combobox, rather than adding another task to the CloseDialog... it shouldn't be  the CloseDialogWithApplyUpdateTheHistoryAndZoomToPointToolbar() routine. ;) 


Just to throw another wrench in the works... have we thought about the cases where either the Point Names are hidden,  an individual point name is hidden, or groups are hidden? You could be zooming to nothing.

 Just a thought... what if when zooming to a point, if it's hidden, which ever of those 3 options is hiding the point name, it's toggled to show the point so the zoom would now have context?  With the added benefit of being able to find points that ARE hidden - which IMO seems even more useful than zooming to a visible point.

slspencer commented:
Hi @alvarocasadoc How is it going?

alvarocasadoc committed:
Made zoom to point dialog combobox editable

alvarocasadoc committed:
Update draft points combobox when doc is updated

alvarocasadoc committed:
Show point name when zooming to it

alvarocasadoc committed:
Removed call to updateZoomToPointComboBox() from showDraftMode() in MainWindow.cpp

alvarocasadoc committed:
Removed duplicate point names in draft points list

alvarocasadoc commented:
Hello everyone. I've been busy for a couple of days, but I'm finally back!

I have tried to address all your comments:
- [x] Made the combobox (in the zoom to point dialog) editable.
- [x] Included `zoomToPointComboBox->blockSignals()` inside the`updateZoomToPointComboBox` function.
- [x] Connected the call to `updateZoomToPointComboBox` to the `doc` signal `CheckLayout` so that any change made to the pattern updates the combobox contents. I have verified that this takes effect when creating, deleting, renaming and updating a point.
- [x] As @DSCaskey suggested, now if a point is hidden, zooming to it now shows its name label. I have thought of including this as an undo command so that is easily undone.
- [x] Removed duplicates from the combobox.

Some loose ends I plan to tie up next:
- [ ] Take care of the alignment problems in the Help>Shortcuts dialog.
- [ ] Show a point name label when Point Names are hidden. Do you think we should toggle the visibility of all points?
- [ ] Show a point name label when its parent group is hidden (I think this should toggle the visibility of the group)
- [ ] Verify that the ? in the Window title bar has disappeared after adding window flags.

DSCaskey commented:
I'll take a look as soon as I can. 

Glad to see that you connected to the doc signal... I knew there was a reason why updating the combobox in the ClosedDialogWithApply() was not correct,  but couldn't think of it at the time.  A user can also change a point name in the Properties Editor which is not handled by the dialog handler.


> I have thought of including this as an undo command so that is easily undone.

Showing the point name? Zooming? Or Both?

>Show a point name label when Point Names are hidden. Do you think we should toggle the visibility of all points?

Keep in mind that there are 2 levels of visibility. Globally toggling all visible points show/ hide, and each individual point. That being said, I think only the zoomed point should be toggled visible, otherwise we can assume a user would then have to go back and hide all the other points. 


>Show a point name label when its parent group is hidden (I think this should toggle the visibility of the group)

You have to toggle the whole group... just like you have to globally toggle all point names visible to see any given one. Not to mention there's more to the groups than just toggling a point name  - the whole tool is shown / hidden.  That also being said... since I plan on updating the Groups in the next few weeks, maybe leave the Groups out  as it's all going to get cleaned up, and prefer to avoid conflicts to resolve.

alvarocasadoc commented:
> Showing the point name? Zooming? Or Both?

Only showing the point name.

> Keep in mind that there are 2 levels of visibility. Globally toggling all visible points show/ hide, and each individual point. That being said, I think only the zoomed point should be toggled visible, otherwise we can assume a user would then have to go back and hide all the other points.

Ok, I will make it toggle only that point

> ... since I plan on updating the Groups in the next few weeks, maybe leave the Groups out as it's all going to get cleaned up, and prefer to avoid conflicts to resolve.

Ok, I will leave the groups visibility as it is now

Thanks!

DSCaskey commented:
@alvarocasadoc 

Everything looks good... except:

>As @DSCaskey suggested, now if a point is hidden, zooming to it now shows its name label. I have thought of including this as an undo command so that is easily undone.

>Show a point name label when Point Names are hidden. Do you think we should toggle the visibility of all points?

I'm getting when Zooming to a point with the name hidden, the name is still hidden. Not sure if this is suppose to be done or is still to do?

 Which after playing around some more it is kind of confusing if the point is hidden.  It zooms,  you're looking for the point, but you can't see it. If we get that working I think we're good to go.

slspencer commented:
@alvarocasadoc  & @DSCaskey 

Is is correct to say the only item remaining is to unhide a point when we zoom to it?

DSCaskey commented:
@slspencer 

Yes, I'd say that... but I'd be open to merging the PR as it stands at this point.

slspencer commented:
Ok, approve the PR in your review & I'll merge it.

DSCaskey commented:
@alvarocasadoc 

Hey Alvaro... I approved the changes you made upto now so Susan can merge them. If you figure out how to show hidden points, you can make a new issue and PR for that addition.

alvarocasadoc commented:
Sure, thank you! Sorry for the absence during these days. I really did not manage to have time for working on this. I will open a new PR when I fix the remaining loose ends 😄
